core/opregion.c

```diff
@@ -38,571 +41,327 @@ void lai_write_opregion(lai_nsnode_t *field, lai_variable_t *source) {
-            if (lai_eval(&bus_number, bbn_handle, &state))
-                lai_panic("could not evaluate _BBN of OperationRegion()");
-            bbn_result = bus_number.integer;
-        }
+    uint64_t seg = 0; // When _SEG is not present, we default to Segment Group 0
+    uint64_t bbn = 0; // When _BBN is not present, we assume PCI bus 0.
+    uint64_t adr = 0; // When _ADR is not present, again, default to zero.
 
-        // Device slot/function is in the _ADR object.
-        lai_nsnode_t *adr_handle = lai_resolve_search(opregion, "_ADR");
-        if (adr_handle) {
-            if (lai_eval(&address_number, adr_handle, &state))
-                lai_panic("could not evaluate _ADR of OperationRegion()");
-            adr_result = address_number.integer;
-        }
+    if (opregion->op_address_space == ACPI_OPREGION_PCI)
+        lai_get_pci_params(opregion, &seg, &bbn, &adr);
 
-       switch (field->fld_flags & 0x0F) {
-            case FIELD_BYTE_ACCESS:
-                if (!laihost_pci_readb) lai_panic("host does not provide PCI access functions");
-                value = laihost_pci_readb((uint16_t)seg_result,
-                                 (uint8_t)bbn_result,
-                                 (uint8_t)(adr_result >> 16) & 0xFF,
-                                 (uint8_t)(adr_result & 0xFF),
-                                 (offset & 0xFFFF) + opregion->op_base);
-                break;
-            case FIELD_WORD_ACCESS:
-                if (!laihost_pci_readw) lai_panic("host does not provide PCI access functions");
-                value = laihost_pci_readw((uint16_t)seg_result,
-                                 (uint8_t)bbn_result,
-                                 (uint8_t)(adr_result >> 16) & 0xFF,
-                                 (uint8_t)(adr_result & 0xFF),
-                                 (offset & 0xFFFE) + opregion->op_base);
-                break;
-            case FIELD_DWORD_ACCESS:
-            case FIELD_ANY_ACCESS:
-                if (!laihost_pci_readd) lai_panic("host does not provide PCI access functions");
-                value = laihost_pci_readd((uint16_t)seg_result,
-                                 (uint8_t)bbn_result,
-                                 (uint8_t)(adr_result >> 16) & 0xFF,
-                                 (uint8_t)(adr_result & 0xFF),
-                                 (offset & 0xFFFC) + opregion->op_base);
-                break;
-            default:
-                lai_panic("undefined field flags 0x%02X: %s", field->fld_flags,
-                          lai_stringify_node_path(field));
-        }
-    } else {
-        lai_panic("undefined opregion address space: %d", opregion->op_address_space);
+    size_t progress = 0;
+    while (progress < field->fld_size) {
+        size_t read_bits = MIN(field->fld_size - progress, access_size - ((bit_offset + progress) % access_size));
+        size_t read_shift = ((field->fld_offset + progress) & (access_size - 1));
```

Comment:
If `bit_offset` is redefined as suggested above, this variable is not necessary anymore.